File: Algorithms/Graph/785. Is Graph Bipartite?.cpp
//Breadth first search
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
		queue<int> q;    
		vector<int> visited(graph.size(), -1);
		for(int i = 0; i < graph.size(); i++)
		{
			if(visited[i] == -1)
			{
				q.emplace(i);
				visited[i] = 0;
				while(!q.empty())
				{
					int node = q.front();
					q.pop();
					for(int& next : graph[node])
					{
						if(visited[next] == visited[node]) return false;
						if(visited[next] == -1)
						{
							q.emplace(next);
							visited[next] = 1 - visited[node];
						}
					}
				}
			}
		}
		return true;
    }
};

//Depth first search
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
		vector<int> visited(graph.size(), -1);
		for(int i = 0; i < graph.size(); i++)
		{
			if(visited[i] == -1)
			{
				visited[i] = 0;
				if(!dfs(graph, visited, i)) return false;	
			}
		}
		return true;
	}
private:
	bool dfs(vector<vector<int>>& graph, vector<int>& visited ,int node) {
		for(int& next : graph[node])
		{
			if(visited[next] == -1)
			{
				visited[next] = 1 - visited[node];
				if(!dfs(graph, visited, next)) return false;
			}
			else if(visited[next] == visited[node])
			{
				return false;
			}
		}
		return true;
	}
};

File: Algorithms/Graph/785. Is Graph Bipartite?.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int[] colors = new int[n];
        for(int i = 0; i < n; ++i) {
            if(colors[i] == 0) {
                colors[i] = 1;
                if(!dfs(graph, colors, i)) return false;
            }
        }
        return true;
    }

    private boolean dfs(int[][] graph, int[] colors, int pos) {
        for(int next : graph[pos]) {
            if(colors[next] == colors[pos]) return false;
            if(colors[next] == 0) {
                colors[next] = 3 - colors[pos];
                if(!dfs(graph, colors, next)) return false;
            }
        }
        return true;
    }
}
